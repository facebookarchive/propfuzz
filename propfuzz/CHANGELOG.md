# Changelog

## [0.0.1] - 2020-07-01
### Added
- Initial release.

[0.0.1]: https://github.com/facebookincubator/cargo-guppy/releases/tag/propfuzz-0.0.1
