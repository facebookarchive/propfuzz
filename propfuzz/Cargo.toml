[package]
name = "propfuzz"
version = "0.0.1"
description = "Combine property-based testing and fuzzing."
repository = "https://github.com/facebookincubator/propfuzz"
authors = ["Rain <rain1@fb.com>"]
license = "MIT OR Apache-2.0"
edition = "2018"
readme = "README.md"
exclude = [
    # Readme template that doesn't need to be included.
    "README.tpl",
]

[badges]
maintenance = { status = "actively-developed" }

[package.metadata.docs.rs]
all-features = true

[dependencies]
proptest = "1.0.0"
propfuzz-macro = { path = "../propfuzz-macro", version = "0.0.1", optional = true }

[dev-dependencies]
pretty_assertions = "1.0.0"
trybuild = "1.0.53"

[features]
default = ["macro"]
macro = ["propfuzz-macro"]

[[test]]
name = "basic"
required-features = ["macro"]

[[test]]
name = "trybuild"
required-features = ["macro"]
