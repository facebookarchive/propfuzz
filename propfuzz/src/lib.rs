// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Rust toolkit to combine property-based testing with fuzzing.
//!
//! For more, see the [`README`](https://github.com/facebookincubator/propfuzz/blob/main/README.md)
//! at the root of the `propfuzz` repository.

pub mod prelude;
pub mod runtime;
pub mod traits;

// Re-export the propfuzz macro -- this is expected to be the primary interface.
#[cfg(feature = "macro")]
pub use propfuzz_macro::propfuzz;

pub use proptest;
