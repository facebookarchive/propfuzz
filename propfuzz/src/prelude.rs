// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Re-exports for the most commonly used APIs of `propfuzz`.
//!
//! This includes:
//! * the `propfuzz` macro from this crate
//! * the entire prelude of `proptest`, so existing tests can be migrated with minimal hassle.
//!
//! ## Examples
//!
//! There's no need to specify `proptest` as a separate dependency, since you can write:
//!
//! ```
//! use propfuzz::prelude::*;
//! use proptest::collection::vec;
//!
//! /// Example test.
//! #[propfuzz]
//! fn test(#[propfuzz(strategy = "vec(any::<u8>(), 0..64)")] v: Vec<u8>) {
//!     // ...
//! }
//! ```

#[doc(no_inline)]
pub use crate::propfuzz;

#[doc(no_inline)]
pub use proptest;
#[doc(no_inline)]
pub use proptest::prelude::*;
