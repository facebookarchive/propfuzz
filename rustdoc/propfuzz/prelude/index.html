<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `prelude` mod in crate `propfuzz`."><meta name="keywords" content="rust, rustlang, rust-lang, prelude"><title>propfuzz::prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../propfuzz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module prelude</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#reexports">Re-exports</a></li></ul></div><p class="location"><a href="../index.html">propfuzz</a></p><script>window.sidebarCurrent = {name: "prelude", ty: "mod", relpath: "../"};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/propfuzz/prelude.rs.html#4-31" title="goto source code">[src]</a></span><span class="in-band">Module <a href="../index.html">propfuzz</a>::<wbr><a class="mod" href="">prelude</a></span></h1><div class="docblock"><p>Re-exports for the most commonly used APIs of <code>propfuzz</code>.</p>
<p>This includes:</p>
<ul>
<li>the <code>propfuzz</code> macro from this crate</li>
<li>the entire prelude of <code>proptest</code>, so existing tests can be migrated with minimal hassle.</li>
</ul>
<h2 id="examples" class="section-header"><a href="#examples">Examples</a></h2>
<p>There's no need to specify <code>proptest</code> as a separate dependency, since you can write:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">propfuzz</span>::<span class="ident">prelude</span>::<span class="kw-2">*</span>;
<span class="kw">use</span> <span class="ident">proptest</span>::<span class="ident">collection</span>::<span class="ident">vec</span>;

<span class="doccomment">/// Example test.</span>
<span class="attribute">#[<span class="ident">propfuzz</span>]</span>
<span class="kw">fn</span> <span class="ident">test</span>(<span class="attribute">#[<span class="ident">propfuzz</span>(<span class="ident">strategy</span> <span class="op">=</span> <span class="string">&quot;vec(any::&lt;u8&gt;(), 0..64)&quot;</span>)]</span> <span class="ident">v</span>: <span class="ident">Vec</span><span class="op">&lt;</span><span class="ident">u8</span><span class="op">&gt;</span>) {
    <span class="comment">// ...</span>
}</pre></div>
</div><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<table><tr><td><code>pub use crate::<a class="attr" href="../../propfuzz/attr.propfuzz.html" title="attr propfuzz::propfuzz">propfuzz</a>;</code></td></tr><tr><td><code>pub use <a class="mod" href="../../proptest/index.html" title="mod proptest">proptest</a>;</code></td></tr><tr><td><code>pub use <a class="macro" href="../../proptest/macro.proptest.html" title="macro proptest::proptest">proptest</a>;</code></td></tr><tr><td><code>pub use proptest::<a class="mod" href="../../proptest/prelude/index.html" title="mod proptest::prelude">prelude</a>::*;</code></td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "propfuzz";</script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>