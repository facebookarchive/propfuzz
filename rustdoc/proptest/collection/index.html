<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `collection` mod in crate `proptest`."><meta name="keywords" content="rust, rustlang, rust-lang, collection"><title>proptest::collection - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../proptest/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module collection</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#functions">Functions</a></li></ul></div><p class="location"><a href="../index.html">proptest</a></p><script>window.sidebarCurrent = {name: "collection", ty: "mod", relpath: "../"};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/proptest/collection.rs.html#10-779" title="goto source code">[src]</a></span><span class="in-band">Module <a href="../index.html">proptest</a>::<wbr><a class="mod" href="">collection</a></span></h1><div class="docblock"><p>Strategies for generating <code>std::collections</code> of values.</p>
</div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.BTreeMapStrategy.html" title="proptest::collection::BTreeMapStrategy struct">BTreeMapStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>BTreeMap</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.BTreeMapValueTree.html" title="proptest::collection::BTreeMapValueTree struct">BTreeMapValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>BTreeMapStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.BTreeSetStrategy.html" title="proptest::collection::BTreeSetStrategy struct">BTreeSetStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>BTreeSet</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.BTreeSetValueTree.html" title="proptest::collection::BTreeSetValueTree struct">BTreeSetValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>BTreeSetStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.BinaryHeapStrategy.html" title="proptest::collection::BinaryHeapStrategy struct">BinaryHeapStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>BinaryHeap</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.BinaryHeapValueTree.html" title="proptest::collection::BinaryHeapValueTree struct">BinaryHeapValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>BinaryHeapStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.HashMapStrategy.html" title="proptest::collection::HashMapStrategy struct">HashMapStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>HashMap</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.HashMapValueTree.html" title="proptest::collection::HashMapValueTree struct">HashMapValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>HashMapStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.HashSetStrategy.html" title="proptest::collection::HashSetStrategy struct">HashSetStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>HashSet</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.HashSetValueTree.html" title="proptest::collection::HashSetValueTree struct">HashSetValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>HashSetStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.LinkedListStrategy.html" title="proptest::collection::LinkedListStrategy struct">LinkedListStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>LinkedList</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.LinkedListValueTree.html" title="proptest::collection::LinkedListValueTree struct">LinkedListValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>LinkedListStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.SizeRange.html" title="proptest::collection::SizeRange struct">SizeRange</a></td><td class="docblock-short"><p>The minimum and maximum range/bounds on the size of a collection.
The interval must form a subset of <code>[0, std::usize::MAX)</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.VecDequeStrategy.html" title="proptest::collection::VecDequeStrategy struct">VecDequeStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>VecDeque</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.VecDequeValueTree.html" title="proptest::collection::VecDequeValueTree struct">VecDequeValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>VecDequeStrategy</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.VecStrategy.html" title="proptest::collection::VecStrategy struct">VecStrategy</a></td><td class="docblock-short"><p>Strategy to create <code>Vec</code>s with a length in a certain range.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.VecValueTree.html" title="proptest::collection::VecValueTree struct">VecValueTree</a></td><td class="docblock-short"><p><code>ValueTree</code> corresponding to <code>VecStrategy</code>.</p>
</td></tr></table><h2 id="functions" class="section-header"><a href="#functions">Functions</a></h2>
<table><tr class="module-item"><td><a class="fn" href="fn.binary_heap.html" title="proptest::collection::binary_heap fn">binary_heap</a></td><td class="docblock-short"><p>Create a strategy to generate <code>BinaryHeap</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.btree_map.html" title="proptest::collection::btree_map fn">btree_map</a></td><td class="docblock-short"><p>Create a strategy to generate <code>BTreeMap</code>s containing keys and values drawn
from <code>key</code> and <code>value</code> respectively, and with a size within the given
range.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.btree_set.html" title="proptest::collection::btree_set fn">btree_set</a></td><td class="docblock-short"><p>Create a strategy to generate <code>BTreeSet</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.hash_map.html" title="proptest::collection::hash_map fn">hash_map</a></td><td class="docblock-short"><p>Create a strategy to generate <code>HashMap</code>s containing keys and values drawn
from <code>key</code> and <code>value</code> respectively, and with a size within the given
range.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.hash_set.html" title="proptest::collection::hash_set fn">hash_set</a></td><td class="docblock-short"><p>Create a strategy to generate <code>HashSet</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.linked_list.html" title="proptest::collection::linked_list fn">linked_list</a></td><td class="docblock-short"><p>Create a strategy to generate <code>LinkedList</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.size_range.html" title="proptest::collection::size_range fn">size_range</a></td><td class="docblock-short"><p>Creates a <code>SizeRange</code> from some value that is convertible into it.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.vec.html" title="proptest::collection::vec fn">vec</a></td><td class="docblock-short"><p>Create a strategy to generate <code>Vec</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.vec_deque.html" title="proptest::collection::vec_deque fn">vec_deque</a></td><td class="docblock-short"><p>Create a strategy to generate <code>VecDeque</code>s containing elements drawn from
<code>element</code> and with a size range given by <code>size</code>.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "proptest";</script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>