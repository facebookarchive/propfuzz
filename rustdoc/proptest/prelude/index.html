<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `prelude` mod in crate `proptest`."><meta name="keywords" content="rust, rustlang, rust-lang, prelude"><title>proptest::prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../proptest/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module prelude</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></div><p class="location"><a href="../index.html">proptest</a></p><script>window.sidebarCurrent = {name: "prelude", ty: "mod", relpath: "../"};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/proptest/prelude.rs.html#10-53" title="goto source code">[src]</a></span><span class="in-band">Module <a href="../index.html">proptest</a>::<wbr><a class="mod" href="">prelude</a></span></h1><div class="docblock"><p>Re-exports the most commonly-needed APIs of proptest.</p>
<p>This module is intended to be wildcard-imported, i.e.,
<code>use proptest::prelude::*;</code>. Note that it re-exports the whole crate itself
under the name <code>prop</code>, so you don't need a separate <code>use proptest;</code> line.</p>
<p>In addition to Proptest's own APIs, this also reexports a small portion of
the <code>rand</code> crate sufficient to easily use <code>prop_perturb</code> and other
functionality that exposes random number generators. Please note that this
is will always be a direct reexport; using these in preference to using the
<code>rand</code> crate directly will not provide insulation from the upcoming
revision to the <code>rand</code> crate.</p>
</div><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<table><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_assert.html" title="macro proptest::prop_assert">prop_assert</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_assert_eq.html" title="macro proptest::prop_assert_eq">prop_assert_eq</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_assert_ne.html" title="macro proptest::prop_assert_ne">prop_assert_ne</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_assume.html" title="macro proptest::prop_assume">prop_assume</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_compose.html" title="macro proptest::prop_compose">prop_compose</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.prop_oneof.html" title="macro proptest::prop_oneof">prop_oneof</a>;</code></td></tr><tr><td><code>pub use crate::<a class="macro" href="../../proptest/macro.proptest.html" title="macro proptest::proptest">proptest</a>;</code></td></tr></table><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="prop/index.html" title="proptest::prelude::prop mod">prop</a></td><td class="docblock-short"><p>Re-exports the entire public API of proptest so that an import of <code>prelude</code>
allows simply writing, for example, <code>prop::num::i32::ANY</code> rather than
<code>proptest::num::i32::ANY</code> plus a separate <code>use proptest;</code>.</p>
</td></tr></table><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.BoxedStrategy.html" title="proptest::prelude::BoxedStrategy struct">BoxedStrategy</a></td><td class="docblock-short"><p>A boxed <code>Strategy</code> trait object as produced by <code>Strategy::boxed()</code>.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.Just.html" title="proptest::prelude::Just struct">Just</a></td><td class="docblock-short"><p>A <code>Strategy</code> which always produces a single value value and never
simplifies.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.ProptestConfig.html" title="proptest::prelude::ProptestConfig struct">ProptestConfig</a></td><td class="docblock-short"><p>Configuration for how a proptest test should be run.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.SBoxedStrategy.html" title="proptest::prelude::SBoxedStrategy struct">SBoxedStrategy</a></td><td class="docblock-short"><p>A boxed <code>Strategy</code> trait object which is also <code>Sync</code> and
<code>Send</code>, as produced by <code>Strategy::sboxed()</code>.</p>
</td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.TestCaseError.html" title="proptest::prelude::TestCaseError enum">TestCaseError</a></td><td class="docblock-short"><p>Errors which can be returned from test cases to indicate non-successful
completion.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.Arbitrary.html" title="proptest::prelude::Arbitrary trait">Arbitrary</a></td><td class="docblock-short"><p>Arbitrary determines a canonical <a href="../strategy/trait.Strategy.html"><code>Strategy</code></a> for the implementing type.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.Rng.html" title="proptest::prelude::Rng trait">Rng</a></td><td class="docblock-short"><p>An automatically-implemented extension trait on <a href="../../proptest/prelude/trait.RngCore.html" title="RngCore"><code>RngCore</code></a> providing high-level
generic methods for sampling values and other convenience methods.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.RngCore.html" title="proptest::prelude::RngCore trait">RngCore</a></td><td class="docblock-short"><p>The core of a random number generator.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.Strategy.html" title="proptest::prelude::Strategy trait">Strategy</a></td><td class="docblock-short"><p>A strategy for producing arbitrary values of a given type.</p>
</td></tr></table><h2 id="functions" class="section-header"><a href="#functions">Functions</a></h2>
<table><tr class="module-item"><td><a class="fn" href="fn.any.html" title="proptest::prelude::any fn">any</a></td><td class="docblock-short"><p>Generates a <a href="../strategy/trait.Strategy.html"><code>Strategy</code></a> producing <a href="trait.Arbitrary.html"><code>Arbitrary</code></a> values of
<code>A</code>. Unlike <a href="fn.arbitrary.html"><code>arbitrary</code></a>, it should be used for being
explicit on what <code>A</code> is. For clarity, this may be a good idea.</p>
</td></tr><tr class="module-item"><td><a class="fn" href="fn.any_with.html" title="proptest::prelude::any_with fn">any_with</a></td><td class="docblock-short"><p>Generates a <a href="../strategy/trait.Strategy.html"><code>Strategy</code></a> producing <a href="trait.Arbitrary.html"><code>Arbitrary</code></a> values of <code>A</code> with the
given configuration arguments passed in <code>args</code>. Unlike <a href="fn.arbitrary_with.html"><code>arbitrary_with</code></a>,
it should be used for being explicit on what <code>A</code> is.
For clarity, this may be a good idea.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "proptest";</script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>