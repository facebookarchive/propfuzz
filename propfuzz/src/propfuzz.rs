// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

use proptest::prelude::*;
use proptest::test_runner::TestRunner;
use std::fmt;

/// Represents a structured fuzz target.
///
/// A trait that implements `Propfuzz` can be used both as a standard property-based test, and
/// as a target for structured, mutation-based fuzzing.
///
/// Structured, mutation-based fuzzers use random byte sequences as a pass-through RNG. In other
/// words, the random byte sequences act as something like a DNA for random values.
pub trait Propfuzz: Send + Sync + fmt::Debug {
    /// Returns the name of this structured fuzz target.
    fn name(&self) -> &'static str;

    /// Returns a description for this structured fuzz target.
    fn description(&self) -> &'static str;

    /// Returns the proptest config for this fuzz target.
    ///
    /// Note that the config is modified for the
    fn proptest_config(&self) -> ProptestConfig {
        ProptestConfig::default()
    }

    /// Executes this test using the given test runner.
    ///
    /// This is where the main body of the test goes.
    fn execute(&self, test_runner: &mut TestRunner);
}
