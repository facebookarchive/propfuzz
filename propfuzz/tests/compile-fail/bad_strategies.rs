// Copyright (c) The propfuzz Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Propfuzz with bad strategies.

use propfuzz::prelude::*;
use proptest::collection::vec;
use std::collections::HashSet;

/// Incorrect strategy format.
#[propfuzz]
fn wrong_format(#[propfuzz(strategy(foo))] _: u8) {}

/// Invalid expression (_ is not a valid Rust identifier).
#[propfuzz]
fn invalid_expr(#[propfuzz(strategy = "_")] _: u8) {}

/// Unknown function name.
#[propfuzz]
fn unknown_name(#[propfuzz(strategy = "xxxxxxx(any::<u64>(), 0..8)")] _: HashSet<u64>) {}

/// Unknown attribute.
#[propfuzz]
fn unknown_attr(#[foo] _: u8) {}

/// Wrong type of strategy.
#[propfuzz]
fn wrong_strategy_type(#[propfuzz(strategy = "vec(any::<u64>(), 0..8)")] _: HashSet<u64>) {}

fn main() {}
