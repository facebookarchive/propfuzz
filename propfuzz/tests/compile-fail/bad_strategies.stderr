error: expected key = value format
  --> $DIR/bad_strategies.rs:12:28
   |
12 | fn wrong_format(#[propfuzz(strategy(foo))] _: u8) {}
   |                            ^^^^^^^^^^^^^

error: expected expression, found reserved identifier `_`
  --> $DIR/bad_strategies.rs:16:39
   |
16 | fn invalid_expr(#[propfuzz(strategy = "_")] _: u8) {}
   |                                       ^^^ expected expression

error: non-#[propfuzz] attributes are not supported
  --> $DIR/bad_strategies.rs:24:17
   |
24 | fn unknown_attr(#[foo] _: u8) {}
   |                 ^^^^^^

error[E0425]: cannot find function `xxxxxxx` in this scope
  --> $DIR/bad_strategies.rs:20:39
   |
20 | fn unknown_name(#[propfuzz(strategy = "xxxxxxx(any::<u64>(), 0..8)")] _: HashSet<u64>) {}
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0308]: mismatched types
  --> $DIR/bad_strategies.rs:27:1
   |
27 | #[propfuzz]
   | ^^^^^^^^^^^
   | |
   | expected struct `HashSet`, found struct `Vec`
   | expected `std::result::Result<(), TestError<(HashSet<u64>,)>>` because of return type
   |
   = note: expected enum `std::result::Result<_, TestError<(HashSet<u64>,)>>`
              found enum `std::result::Result<_, TestError<(Vec<u64>,)>>`
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
